File: client/src/utils/tigertagCodec.ts
/**
 * TigerTag NTAG213 binary encoder/decoder for browser-side NFC.
 *
 * Implements the TigerTag Maker format (big-endian) based on the TigerTag
 * RFID Guide specification.
 *
 * Produces 144-byte binary payloads compatible with NTAG213 user memory (pages 4-39).
 */

import { ISpool } from "../pages/spools/model";

/** TigerTag Maker v1.0 magic number (from id_version.json API — the README hex values are incorrect). */
export const TIGERTAG_MAKER_V1 = 0x5bf59264;

/** TigerTag+ v1.0 magic number — offline + cloud sync. */
export const TIGERTAG_PRO_V1 = 0xbc0fcb97;

/** Check if a magic number identifies a valid TigerTag (Maker or Pro/+). */
export function isTigerTag(magic: number): boolean {
  return magic === TIGERTAG_MAKER_V1 || magic === TIGERTAG_PRO_V1;
}

/** Seconds between Unix epoch (1970-01-01) and TigerTag epoch (2000-01-01). */
export const TIGERTAG_EPOCH_OFFSET = 946684800;

/** NTAG213 user memory size in bytes. */
const NTAG213_USER_BYTES = 144;

/** TigerTag Maker header size in bytes. */
const HEADER_SIZE = 36;

/** Maximum user message length in bytes. */
const USER_MESSAGE_SIZE = 28;

/** User message offset in the 144-byte payload. */
const USER_MESSAGE_OFFSET = 58;

/** Emoji offset. */
const EMOJI_OFFSET = 54;

/** Bed temp offset (2 × uint8 after the 36-byte header). */
const BED_TEMP_OFFSET = 36;

/** TigerTag filament type ID. */
const TIGERTAG_TYPE_FILAMENT = 142;

export interface TigerTagBinaryData {
  id_tigertag: number;
  id_product: number;
  id_material: number;
  id_diameter: number;
  id_aspect: number;
  id_type: number;
  id_brand: number;
  color_r: number;
  color_g: number;
  color_b: number;
  color_a: number;
  weight: number;
  nozzle_temp: number;
  nozzle_temp_max: number;
  bed_temp: number;
  bed_temp_max: number;
  drying_temp: number;
  drying_duration: number;
  timestamp: number;
  emoji: number;
  user_message: string;
}

// TigerTag binary format (big-endian, 36-byte header):
//
// Offset  Size  Field
// 0       4     id_tigertag (uint32 BE)
// 4       4     id_product (uint32 BE)
// 8       2     id_material (uint16 BE)
// 10      1     id_aspect_1 (uint8)
// 11      1     id_aspect_2 (uint8)
// 12      1     id_type (uint8)
// 13      1     id_diameter (uint8)
// 14      2     id_brand (uint16 BE)
// 16      4     color_rgba (uint32 BE: R<<24 | G<<16 | B<<8 | A)
// 20      4     weight_unit (uint32 BE: weight<<8 | unit_id)
// 24      2     nozzle_temp_min (uint16 BE)
// 26      2     nozzle_temp_max (uint16 BE)
// 28      1     drying_temp (uint8)
// 29      1     drying_time (uint8, hours)
// 30      2     reserved (uint16)
// 32      4     timestamp (uint32 BE)
//
// After header:
// 36      1     bed_temp_min (uint8)
// 37      1     bed_temp_max (uint8)
// 38-53   16    reserved
// 54      4     emoji (uint32 BE)
// 58      28    user_message (UTF-8)
// 86-143  58    signature / reserved

/**
 * Decode a TigerTag binary payload into structured data.
 * Uses big-endian format as per the TigerTag RFID Guide specification.
 */
export function decodeTigerTag(buf: ArrayBuffer): TigerTagBinaryData {
  if (buf.byteLength < HEADER_SIZE) {
    throw new Error(`Data too short: expected at least ${HEADER_SIZE} bytes, got ${buf.byteLength}`);
  }

  const view = new DataView(buf);
  const BE = false; // DataView littleEndian=false → big-endian

  let offset = 0;
  const id_tigertag = view.getUint32(offset, BE); offset += 4;
  const id_product = view.getUint32(offset, BE); offset += 4;
  const id_material = view.getUint16(offset, BE); offset += 2;
  const id_aspect = view.getUint8(offset); offset += 1;  // aspect_1
  offset += 1; // aspect_2, skip
  const id_type = view.getUint8(offset); offset += 1;
  const id_diameter = view.getUint8(offset); offset += 1;
  const id_brand = view.getUint16(offset, BE); offset += 2;

  // Color packed as uint32 BE: R<<24 | G<<16 | B<<8 | A
  const colorVal = view.getUint32(offset, BE); offset += 4;
  const color_r = (colorVal >>> 24) & 0xff;
  const color_g = (colorVal >>> 16) & 0xff;
  const color_b = (colorVal >>> 8) & 0xff;
  const color_a = colorVal & 0xff;

  // Weight + unit packed as uint32 BE: weight in upper 24 bits, unit in lower 8
  const weightUnit = view.getUint32(offset, BE); offset += 4;
  const weight = (weightUnit >>> 8) & 0xffffff;

  const nozzle_temp = view.getUint16(offset, BE); offset += 2;
  const nozzle_temp_max = view.getUint16(offset, BE); offset += 2;
  const drying_temp = view.getUint8(offset); offset += 1;
  const drying_duration = view.getUint8(offset); offset += 1;
  offset += 2; // reserved
  const timestamp = view.getUint32(offset, BE); offset += 4;

  // Bed temp at offset 36-37
  let bed_temp = 0;
  let bed_temp_max = 0;
  if (buf.byteLength > BED_TEMP_OFFSET + 1) {
    bed_temp = view.getUint8(BED_TEMP_OFFSET);
    bed_temp_max = view.getUint8(BED_TEMP_OFFSET + 1);
  }

  // Emoji at offset 54
  let emoji = 0;
  if (buf.byteLength >= EMOJI_OFFSET + 4) {
    emoji = view.getUint32(EMOJI_OFFSET, BE);
  }

  // User message at offset 58
  let user_message = "";
  if (buf.byteLength >= USER_MESSAGE_OFFSET + USER_MESSAGE_SIZE) {
    const msgSlice = new Uint8Array(buf, USER_MESSAGE_OFFSET, USER_MESSAGE_SIZE);
    const nullIdx = msgSlice.indexOf(0);
    const msgBytes = nullIdx >= 0 ? msgSlice.slice(0, nullIdx) : msgSlice;
    user_message = new TextDecoder().decode(msgBytes);
  }

  return {
    id_tigertag, id_product, id_material, id_diameter, id_aspect, id_type, id_brand,
    color_r, color_g, color_b, color_a, weight,
    nozzle_temp, nozzle_temp_max, bed_temp, bed_temp_max,
    drying_temp, drying_duration, timestamp, emoji, user_message,
  };
}

/**
 * Encode TigerTag data into a 144-byte ArrayBuffer for NTAG213 user memory.
 * Uses big-endian format as per the TigerTag RFID Guide specification.
 */
export function encodeTigerTag(data: TigerTagBinaryData): ArrayBuffer {
  const buf = new ArrayBuffer(NTAG213_USER_BYTES);
  const view = new DataView(buf);
  const BE = false; // DataView littleEndian=false → big-endian

  let offset = 0;

  view.setUint32(offset, data.id_tigertag >>> 0, BE); offset += 4;
  view.setUint32(offset, data.id_product >>> 0, BE); offset += 4;
  view.setUint16(offset, data.id_material & 0xffff, BE); offset += 2;
  view.setUint8(offset, data.id_aspect & 0xff); offset += 1; // aspect_1
  view.setUint8(offset, 0); offset += 1; // aspect_2
  view.setUint8(offset, data.id_type & 0xff); offset += 1;
  view.setUint8(offset, data.id_diameter & 0xff); offset += 1;
  view.setUint16(offset, data.id_brand & 0xffff, BE); offset += 2;

  // Color packed as uint32 BE: R<<24 | G<<16 | B<<8 | A
  const colorVal = ((data.color_r & 0xff) << 24) | ((data.color_g & 0xff) << 16) |
                   ((data.color_b & 0xff) << 8) | (data.color_a & 0xff);
  view.setUint32(offset, colorVal >>> 0, BE); offset += 4;

  // Weight + unit BE: weight<<8 | unit (unit=1 for grams)
  const weightUnit = ((data.weight & 0xffffff) << 8) | 1;
  view.setUint32(offset, weightUnit >>> 0, BE); offset += 4;

  view.setUint16(offset, data.nozzle_temp & 0xffff, BE); offset += 2;
  view.setUint16(offset, (data.nozzle_temp_max || 0) & 0xffff, BE); offset += 2;
  view.setUint8(offset, data.drying_temp & 0xff); offset += 1;
  view.setUint8(offset, data.drying_duration & 0xff); offset += 1;
  view.setUint16(offset, 0, BE); offset += 2; // reserved
  view.setUint32(offset, data.timestamp >>> 0, BE); offset += 4;

  // Bed temp at offset 36-37
  view.setUint8(BED_TEMP_OFFSET, data.bed_temp & 0xff);
  view.setUint8(BED_TEMP_OFFSET + 1, (data.bed_temp_max || 0) & 0xff);

  // Emoji at offset 54
  view.setUint32(EMOJI_OFFSET, data.emoji >>> 0, BE);

  // User message at offset 58 (28 bytes, null-padded)
  const encoder = new TextEncoder();
  const msgBytes = encoder.encode(data.user_message).slice(0, USER_MESSAGE_SIZE);
  const uint8 = new Uint8Array(buf);
  uint8.set(msgBytes, USER_MESSAGE_OFFSET);

  return buf;
}

/**
 * Map an ISpool to TigerTag binary data, mirroring the Python map_spool_to_tigertag().
 */
export function mapSpoolToTigerTag(spool: ISpool, userMessage: string = ""): TigerTagBinaryData {
  const filament = spool.filament;

  const data: TigerTagBinaryData = {
    id_tigertag: TIGERTAG_MAKER_V1,
    id_product: 0,
    id_material: 0,
    id_diameter: 0,
    id_aspect: 0,
    id_type: TIGERTAG_TYPE_FILAMENT,
    id_brand: 0,
    color_r: 0,
    color_g: 0,
    color_b: 0,
    color_a: 255,
    weight: 0,
    nozzle_temp: 0,
    nozzle_temp_max: 0,
    bed_temp: 0,
    bed_temp_max: 0,
    drying_temp: 0,
    drying_duration: 0,
    timestamp: Math.floor(Date.now() / 1000) - TIGERTAG_EPOCH_OFFSET,
    emoji: 0,
    user_message: userMessage,
  };

  // Set product ID: use TigerTag product ID if available, otherwise use spool ID
  if (filament.external_id?.startsWith("tigertag_")) {
    const parsed = parseInt(filament.external_id.split("_")[1], 10);
    data.id_product = !isNaN(parsed) ? parsed : spool.id;
  } else {
    data.id_product = spool.id;
  }

  // Diameter
  if (filament.diameter) {
    if (Math.abs(filament.diameter - 1.75) < 0.1) {
      data.id_diameter = 1;
    } else if (Math.abs(filament.diameter - 2.85) < 0.1) {
      data.id_diameter = 2;
    }
  }

  // Color
  if (filament.color_hex) {
    const hex = filament.color_hex.replace(/^#/, "");
    if (hex.length >= 6) {
      data.color_r = parseInt(hex.substring(0, 2), 16);
      data.color_g = parseInt(hex.substring(2, 4), 16);
      data.color_b = parseInt(hex.substring(4, 6), 16);
      if (hex.length >= 8) {
        data.color_a = parseInt(hex.substring(6, 8), 16);
      }
    }
  }

  // Weight
  if (filament.weight) {
    data.weight = Math.round(filament.weight);
  }

  // Temperatures
  if (filament.settings_extruder_temp) {
    data.nozzle_temp = filament.settings_extruder_temp;
  }
  if (filament.settings_bed_temp) {
    data.bed_temp = filament.settings_bed_temp;
  }

  return data;
}

File: spoolman/tigertag_codec.py
"""TigerTag NTAG213 binary encoder/decoder.

Implements the TigerTag Maker format for encoding/decoding filament data
to/from NTAG213 NFC chips (144 bytes user memory, pages 4-39).

Based on the TigerTag RFID Guide specification.
"""

import logging
import struct
from dataclasses import dataclass
from typing import Optional

logger = logging.getLogger(__name__)

# NTAG213 has 144 bytes of user memory (pages 4-39, 36 pages x 4 bytes)
NTAG213_USER_BYTES = 144

# TigerTag magic numbers (from id_version.json API database — the README hex values are incorrect)
TIGERTAG_MAKER_V1 = 0x5BF59264  # TigerTag Maker v1.0 — offline only (decimal 1542820452)
TIGERTAG_PRO_V1 = 0xBC0FCB97  # TigerTag+ v1.0 — offline + cloud sync (decimal 3155151767)
TIGERTAG_INIT = 0x6C41A2E1  # TigerTag Init — blank/uninitialized (decimal 1816240865)

# All valid TigerTag magic numbers (for detection)
TIGERTAG_MAGIC_NUMBERS = {TIGERTAG_MAKER_V1, TIGERTAG_PRO_V1}


def is_tigertag(magic: int) -> bool:
    """Check if a magic number identifies a valid TigerTag (Maker or Pro/+)."""
    return magic in TIGERTAG_MAGIC_NUMBERS


@dataclass
class TigerTagData:
    """Represents all data fields stored on a TigerTag NFC chip."""

    # Core identifiers
    id_tigertag: int = 0  # 4 bytes - magic number / format identifier
    id_product: int = 0  # 4 bytes - product/filament ID (0xFFFFFFFF = Factory)
    id_material: int = 0  # 2 bytes - material type ID
    id_diameter: int = 0  # 1 byte - diameter ID (TigerTag numbering)
    id_aspect: int = 0  # 1 byte - aspect/finish ID
    id_type: int = 0  # 1 byte - type ID (142=filament)
    id_brand: int = 0  # 2 bytes - brand/manufacturer ID

    # Color as RGBA
    color_r: int = 0  # 1 byte
    color_g: int = 0  # 1 byte
    color_b: int = 0  # 1 byte
    color_a: int = 255  # 1 byte

    # Spool properties
    weight: int = 0  # net weight in grams

    # Temperature settings
    nozzle_temp: int = 0  # nozzle temp min in °C
    nozzle_temp_max: int = 0  # nozzle temp max in °C
    bed_temp: int = 0  # bed temp min in °C
    bed_temp_max: int = 0  # bed temp max in °C
    drying_temp: int = 0  # drying temp in °C
    drying_duration: int = 0  # drying time in hours

    # Metadata
    timestamp: int = 0  # 4 bytes - seconds since 2000-01-01 GMT
    emoji: int = 0  # 4 bytes - emoji codepoint

    # User message - 28 bytes UTF-8
    user_message: str = ""

    @property
    def color_hex(self) -> str:
        """Get color as hex string (without alpha)."""
        return f"{self.color_r:02x}{self.color_g:02x}{self.color_b:02x}"

    @color_hex.setter
    def color_hex(self, value: str) -> None:
        """Set color from hex string."""
        value = value.lstrip("#")
        if len(value) == 6:
            self.color_r = int(value[0:2], 16)
            self.color_g = int(value[2:4], 16)
            self.color_b = int(value[4:6], 16)
        elif len(value) == 8:
            self.color_r = int(value[0:2], 16)
            self.color_g = int(value[2:4], 16)
            self.color_b = int(value[4:6], 16)
            self.color_a = int(value[6:8], 16)

    @property
    def diameter_mm(self) -> float:
        """Get diameter in mm from the diameter ID."""
        if self.id_diameter == 1:
            return 1.75
        if self.id_diameter == 2:
            return 2.85
        # TigerTag uses larger IDs — look up common ones
        if self.id_diameter == 56:
            return 1.75
        if self.id_diameter == 57:
            return 2.85
        return 0.0


# TigerTag binary format (big-endian, 36-byte header):
#
# Offset  Size  Format  Field
# 0       4     >I      id_tigertag
# 4       4     >I      id_product
# 8       2     >H      id_material
# 10      1     B       id_aspect_1
# 11      1     B       id_aspect_2
# 12      1     B       id_type
# 13      1     B       id_diameter
# 14      2     >H      id_brand
# 16      4     >I      color_rgba (R<<24 | G<<16 | B<<8 | A)
# 20      4     >I      weight_unit (weight<<8 | unit_id)
# 24      2     >H      nozzle_temp_min
# 26      2     >H      nozzle_temp_max
# 28      1     B       drying_temp
# 29      1     B       drying_time (hours)
# 30      2     >H      reserved (0)
# 32      4     >I      timestamp
#
# After header:
# 36      1     B       bed_temp_min
# 37      1     B       bed_temp_max
# 38-53   16    -       reserved
# 54      4     >I      emoji
# 58      28    UTF-8   user_message
# 86-143  58    -       signature / reserved

_HEADER_FMT = ">II HBB BBH I I HH BBH I"
_HEADER_SIZE = struct.calcsize(_HEADER_FMT)  # 36 bytes
_USER_MESSAGE_SIZE = 28
_USER_MESSAGE_OFFSET = 58
_EMOJI_OFFSET = 54
_BED_TEMP_OFFSET = 36


def decode_ntag213(raw_bytes: bytes) -> TigerTagData:
    """Decode raw NTAG213 user memory bytes into TigerTagData.

    Args:
        raw_bytes: The raw bytes from NTAG213 pages 4-39 (up to 144 bytes).

    Returns:
        TigerTagData: The decoded tag data.

    Raises:
        ValueError: If the data is too short to decode.

    """
    if len(raw_bytes) < _HEADER_SIZE:
        raise ValueError(f"Data too short: expected at least {_HEADER_SIZE} bytes, got {len(raw_bytes)}")

    # Debug: dump raw bytes for diagnosing format issues
    hex_dump = " ".join(f"{b:02x}" for b in raw_bytes[:68])
    logger.info("TigerTag raw bytes (first 68): %s", hex_dump)

    values = struct.unpack_from(_HEADER_FMT, raw_bytes, 0)

    # Unpack color from RGBA uint32: value = R<<24 | G<<16 | B<<8 | A
    color_val = values[8]
    color_r = (color_val >> 24) & 0xFF
    color_g = (color_val >> 16) & 0xFF
    color_b = (color_val >> 8) & 0xFF
    color_a = color_val & 0xFF

    # Unpack weight + unit: weight in upper 24 bits, unit_id in lower 8
    weight_unit = values[9]
    weight = (weight_unit >> 8) & 0xFFFFFF

    data = TigerTagData(
        id_tigertag=values[0],
        id_product=values[1],
        id_material=values[2],
        id_aspect=values[3],  # aspect_1
        # values[4] = aspect_2, ignored
        id_type=values[5],
        id_diameter=values[6],
        id_brand=values[7],
        color_r=color_r,
        color_g=color_g,
        color_b=color_b,
        color_a=color_a,
        weight=weight,
        nozzle_temp=values[10],
        nozzle_temp_max=values[11],
        drying_temp=values[12],
        drying_duration=values[13],
        # values[14] = reserved
        timestamp=values[15],
    )

    # Read bed temp (uint8 each) at offset 36-37
    if len(raw_bytes) > _BED_TEMP_OFFSET + 1:
        data.bed_temp = raw_bytes[_BED_TEMP_OFFSET]
        data.bed_temp_max = raw_bytes[_BED_TEMP_OFFSET + 1]

    # Read emoji at offset 54
    if len(raw_bytes) >= _EMOJI_OFFSET + 4:
        data.emoji = struct.unpack_from(">I", raw_bytes, _EMOJI_OFFSET)[0]

    # Read user message at offset 58
    if len(raw_bytes) >= _USER_MESSAGE_OFFSET + _USER_MESSAGE_SIZE:
        msg_bytes = raw_bytes[_USER_MESSAGE_OFFSET : _USER_MESSAGE_OFFSET + _USER_MESSAGE_SIZE]
        null_idx = msg_bytes.find(b"\x00")
        if null_idx >= 0:
            msg_bytes = msg_bytes[:null_idx]
        data.user_message = msg_bytes.decode("utf-8", errors="replace")

    return data


def encode_ntag213(data: TigerTagData) -> bytes:
    """Encode TigerTagData into raw bytes for NTAG213 user memory.

    Returns:
        bytes: 144 bytes to write to NTAG213 pages 4-39.

    """
    # Pack color as RGBA uint32: R<<24 | G<<16 | B<<8 | A
    color_val = (data.color_r << 24) | (data.color_g << 16) | (data.color_b << 8) | data.color_a

    # Pack weight + unit_id: weight<<8 | unit (unit=1 for grams)
    weight_unit = ((data.weight & 0xFFFFFF) << 8) | 1

    header = struct.pack(
        _HEADER_FMT,
        data.id_tigertag,
        data.id_product,
        data.id_material,
        data.id_aspect,  # aspect_1
        0,  # aspect_2
        data.id_type,
        data.id_diameter,
        data.id_brand,
        color_val,
        weight_unit,
        data.nozzle_temp,
        data.nozzle_temp_max,
        data.drying_temp,
        data.drying_duration,
        0,  # reserved
        data.timestamp,
    )

    # Build full 144-byte payload
    payload = bytearray(NTAG213_USER_BYTES)
    payload[: len(header)] = header

    # Bed temp at offset 36-37
    payload[_BED_TEMP_OFFSET] = data.bed_temp & 0xFF
    payload[_BED_TEMP_OFFSET + 1] = data.bed_temp_max & 0xFF

    # Emoji at offset 54
    struct.pack_into(">I", payload, _EMOJI_OFFSET, data.emoji)

    # User message at offset 58 (28 bytes, null-padded)
    msg_bytes = data.user_message.encode("utf-8")[:_USER_MESSAGE_SIZE]
    payload[_USER_MESSAGE_OFFSET : _USER_MESSAGE_OFFSET + len(msg_bytes)] = msg_bytes

    return bytes(payload)
